File: tzapi4android/src/main/java/com/tianzunchina/android/api/utils/UnitConverter.java
package com.tianzunchina.android.api.utils;

import android.app.Application;
import android.util.Base64;

import com.tianzunchina.android.api.control.TZApplication;
import com.tianzunchina.android.api.utils.model.LatLon;
import com.tianzunchina.android.api.utils.trans.Trans;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;

/**
 * 常用数据类型转换器
 * CraetTime 2016-3-7
 * @author SunLiang
 */
public class UnitConverter {
    private static UnitConverter converter;
    private Application app = TZApplication.getInstence();

    public static UnitConverter getInstence(){
        if(converter == null){
            converter = new UnitConverter();
        }
        return converter;
    }

    private UnitConverter(){
    }

    /**
     * 把bytes转换为base64
     */
    public String byte2base64(byte[] bytes) {
        if (bytes == null || bytes.length == 0) {
            return null;
        }
        String str = Base64.encodeToString(bytes, Base64.DEFAULT);
        return str;
    }
    /**
     * 把base64转换为bytes
     */
    public byte[] base642byte(String str) {
        if (str == null || str == "") {
            return null;
        }
        return Base64.decode(str, Base64.DEFAULT);
    }

    public byte[] file2byte(File file) {
        byte[] bytes = null;
        try {
            if (file != null && file.exists()) {
                InputStream is = new FileInputStream(file);
                int length = (int) file.length();
                if (length > Integer.MAX_VALUE) {
                    System.err.println("this file is max ");
                }
                bytes = new byte[length];
                int offset = 0;
                int numRead = 0;
                while (offset < bytes.length
                        && (numRead = is.read(bytes, offset, bytes.length
                        - offset)) >= 0) {
                    offset += numRead;
                }
                if (offset < bytes.length) {
                }
                is.close();
            }
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return bytes;
    }

    /**
     * 由dp至px进行转换
     * 例如 dip2px(1) 将得到1dp对应的px值
     * @param dpValue
     * @return
     */
    public int dip2px(float dpValue) {
        final float scale = app.getResources()
                .getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    /**
     * 由dp至px进行转换
     * 例如 px2dip(1) 将得到1px对应的dp值
     * @param pxValue
     * @return
     */
    public int px2dip(float pxValue) {
        final float scale = app.getResources()
                .getDisplayMetrics().density;
        return (int) (pxValue / scale + 0.5f);
    }

    /**
     * 84坐标系转大地2000
     * @param lon 经度
     * @param lat 纬度
     * @return
     */
    public LatLon WGS84ToCGCS2000(double lon, double lat){
        if(lon == 0){
            return new LatLon(0, 0);
        }
        Trans trans = new Trans(1);
        double x = lon, y = lat;
        double[] coords = trans.doTrans(x, y);
        x = coords[0];
        y = coords[1] - 50;
        return new LatLon(y, x);
    }
}

File: tzapi4android/src/main/java/com/tianzunchina/android/api/utils/UnitConverter.kt
package com.tianzunchina.android.api.utils

import android.util.Base64
import com.tianzunchina.android.api.control.TZApplication

import com.tianzunchina.android.api.utils.model.LatLon
import com.tianzunchina.android.api.utils.trans.Trans

import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.IOException

/**
 * 常用数据类型转换器
 * CraetTime 2016-3-7
 * @author SunLiang
 */
object  UnitConverter{

	/**
	* 把bytes转换为base64
	 */
    fun byte2base64(bytes: ByteArray?): String? {
        if (bytes == null || bytes.size == 0) {
            return null
        }
        val str = Base64.encodeToString(bytes, Base64.DEFAULT)
        return str
    }

	/**
	 * 把base64转换为bytes
	 */
    fun base642byte(str: String?): ByteArray? {
        if (str == null || str === "") {
            return null
        }
        return Base64.decode(str, Base64.DEFAULT)
    }

	/**
	*
	 */
    fun file2byte(file: File?): ByteArray {
        var bytes: ByteArray? = null
        try {
            if (file != null && file.exists()) {
                val `is` = FileInputStream(file)
                val length = file.length().toInt()
                if (length > Integer.MAX_VALUE) {
                    System.err.println("this file is max ")
                }
                bytes = ByteArray(length)
                var offset = 0
                var numRead = 0
                while (offset < bytes.size) {
					numRead = `is`.read(bytes, offset, bytes.size - offset)
					if(numRead >= 0){
						offset += numRead
					} else {
						break
					}
                }
                if (offset < bytes.size) {
                }
                `is`.close()
            }
        } catch (e: FileNotFoundException) {
            e.printStackTrace()
        } catch (e: IOException) {
            e.printStackTrace()
        }
        return bytes!!
    }

    /**
     * 由dp至px进行转换
     * 例如 dip2px(1) 将得到1dp对应的px值
     * @param dpValue
     * *
     * @return
     */
    fun dip2px(dpValue: Float): Int {
        val scale = TZApplication.instance!!.resources.displayMetrics.density
        return (dpValue * scale + 0.5f).toInt()
    }

    /**
     * 由dp至px进行转换
     * 例如 px2dip(1) 将得到1px对应的dp值
     * @param pxValue
     * *
     * @return
     */
    fun px2dip(pxValue: Float): Int {
        val scale = TZApplication.instance!!.resources.displayMetrics.density
        return (pxValue / scale + 0.5f).toInt()
    }

    /**
     * 84坐标系转大地2000
     * @param lon 经度
     * *
     * @param lat 纬度
     * *
     * @return
     */
    fun WGS84ToCGCS2000(lon: Double, lat: Double): LatLon {
        if (lon == 0.0) {
            return LatLon(0.0, 0.0)
        }
        val trans = Trans(1)
        var x = lon
        var y = lat
        val coords = trans.doTrans(x, y)
        x = coords[0]
        y = coords[1] - 50
        return LatLon(y, x)
    }
}
